adjust harmonise function

florence-bockting opened the pull request:
## Description

## Checklist

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Changelog item added to `changelog/`

florence-bockting committed:
adjust harmonise function

florence-bockting committed:
WP: restructuring code

florence-bockting committed:
refactoring

znichollscr committed:
Add basic test of product of splines

znichollscr committed:
Pass spline tests

znichollscr committed:
Pass spline tests

florence-bockting committed:
Merge pull request #14 from climate-resource/reset-fb-zn

Add splines for the cosine-decay

florence-bockting committed:
cleaning up code

florence-bockting committed:
harmonised_spline in harmonise function have to be evaluate point-wise

florence-bockting committed:
add changelog and fix formatting issue in docs

florence-bockting committed:
added comments to code for better understanding

znichollscr commented on TODO.md:
```diff
@@ -4,12 +4,12 @@
+      all it needs to do is call `timeseries_to_spline` really (check)
 
 - fix up the tests in `test_harmonisation_integration`
 - any other general clean up
```
Just to check, have we done these last two points? If yes, we can (and should) delete this file

znichollscr commented on changelog/11.breaking.md:
```diff
@@ -0,0 +1,3 @@
+Reduced number of functions to a minimum of relevant functions and created new modules to group functions thematically
+instead of having them all together in one module (as before in the utils.py module).
+Created some additional unit tests and updated the tutorial in the documentation.
```
Nice. The changelog is aimed at users. Now, we obviously don't have any users apart from ourselves, so we can write whatever we want. However, to build good habits, I'd suggest considering whether you would write this differently if you were aiming this at someone who wasn't a developer. For example:

- which functions did we delete?
- where should users look to see the new structure (in an ideal case, we'd list the modules, in this case I think it's fine to say, "The project is young, look in `src/gradient_aware_harmonisation` to see the new structure" or something similar)

To keep the messages a bit clearer, the note about the tutorial can go in a file called `changelog/11.docs.md`. The note about the tests can go in a file called `changelog/11.trivial.md`.

znichollscr replied:
```suggestion
Reduced the number of functions to the minimum relevant functions and created new modules to group functions thematically
instead of having them all together in one module (as before in the `utils.py` module).
```

> Created some additional unit tests and updated the tutorial in the documentation.

Can go in separate CHANGELOG files (details here: https://github.com/climate-resource/gradient-aware-harmonisation/pull/11/files#r2017054396)

znichollscr commented on src/gradient_aware_harmonisation/__init__.py:
```diff
@@ -18,10 +22,11 @@ def harmonise(  # noqa: PLR0913
-    # convergence_function: Callable[[Spline, Spline], Spline] | None = None,
+    convergence_timeseries: Optional[Timeseries] = None,
+    convergence_time: Optional[Union[int, float]] = None,
+    get_harmonised_spline: GetHarmonisedSplineLike = (
```
The docstring will need to be updated too (currently it refers to `convergence_method`, which no longer exists)

znichollscr commented on src/gradient_aware_harmonisation/__init__.py:
```diff
@@ -65,23 +70,71 @@ def harmonise(  # noqa: PLR0913
-    harmonised_timeseries = harmonised_spline.to_timeseries(
-        time_axis=harmonisee_timeseries.time_axis
+    # convert harmonised spline to timeseries
+    # TODO: passing an array to harmonised_spline leads to a very strange
```
Hmm surprising, I'll take a look

znichollscr commented on src/gradient_aware_harmonisation/__init__.py:
```diff
@@ -65,23 +70,71 @@ def harmonise(  # noqa: PLR0913
+    if convergence_timeseries is None:
+        convergence_timeseries = harmonisee_timeseries
+
+    # convert timeseries to splines
+    target_spline = target_timeseries.to_spline()
+    harmonisee_spline = harmonisee_timeseries.to_spline()
+    convergence_spline = convergence_timeseries.to_spline()
```
Let's switch to using `harmonise_splines` here rather than having duplicate code

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+    Is this helper being applied to the convergence spline?
+    """
+
+    # domain: ClassVar[list[float, float]] = [
+    #     np.finfo(np.float64).tiny,
+    #     np.finfo(np.float64).max,
+    # ]
+    # """Domain of spline"""
```
Note to self: take a look at cleaning this up

znichollscr replied:
One for a future MR

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+            # the weighted sum for computing the harmonised AND converged
+            # function has the form: "gamma * harmonised + (1-gamma) * convergence"
+            # depending on which product we want to compute (LHS or RHS of sum),
+            # we need gamma or 1-gamma, therefore we include the following condition
```
Nice clarification

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+
+            return decay_derivative(x)
+
+        # apply decay function only to values that lie between harmonisation
+        # time and convergence-time
```
Something to consider (there isn't a right answer, so just saying this as something to think about): I generally don't have comments like this. The reason is that either a) the code is clear enough that the code tells you what is going on and the comment isn't needed or b) the code isn't clear enough and should be made clearer. There are some exceptions, but I wonder if there is another way to write the code that would make this comment redundant (e.g. rename `x_decay` to `indexes_in_which_to_use_decay`).

florence-bockting replied:
Agreed.
I find it sometimes a bit difficult to get a notion of "what is obvious" after having intensively worked on a project as a lot of things seem to be obvious in this moment but become less obvious after having touched the project a month and looking again at it.

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+            apply_to_convergence=self.apply_to_convergence,
+        )
+
+    def antiderivative(self) -> SumOfSplines:
```
```suggestion
    def antiderivative(self) -> CosineDecaySplineHelperDerivative:
```

It's not used, but probably clearer like this

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+
+
+@define
+class CosineDecaySplineHelperDerivative:
```
nit: I'd be tempted to define this class after `CosineDecaySpline` (just easier to understand that way around I think)

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+            apply_to_convergence=self.apply_to_convergence,
+        )
+
+    def antiderivative(self) -> SumOfSplines:
```
```suggestion
    def antiderivative(self) -> CosineDecaySplineHelper:
```

As above, might be clearer

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+    convergence_spline: Spline,
+) -> SumOfSplines:
+    """
+    Generate the harmonised spline
```
```suggestion
    Generate the harmonised spline based on a cosine-decay
```

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+    # the target-spline at the harmonisation time (wrt to zero-and
+    # first order derivative). Then we use a decay function to let
+    # the harmonised spline converge to the convergenve-spline (by
+    # default: harmonisee). This decay function has the form of a
```
```suggestion
    # the harmonised spline converge to the convergence-spline. 
    # This decay function has the form of a
```

The harmonisee isn't defined within this context, so I would try and avoid referring to it

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+    # weight * harmonised_spline + (1-weight) * convergence_spline
+    # With weights decaying from 1 to 0 whereby the decay trajectory
+    # is determined by the corresponding approach (e.g. cosine,
+    # polynomial, etc.)
```
```suggestion
    # is determined by the cosine decay.
```

Given that this function only allows us to do cosine-decay, I wouldn't mention polynomials (we can implement that next week).

znichollscr commented on tests/integration/test_harmonise_splines_integration.py:
```diff
@@ -108,7 +83,7 @@ def test_target_and_harmonisee_equal(convergence_time, harmonisation_time):
-        convergence_spline=target,
+        converge_to=target,
         convergence_time=convergence_time,
         # TODO: think about convergence method
```
```suggestion
```

I think we can delete this here and deal with it in a future merge requeset

znichollscr commented on tests/integration/test_harmonise_splines_integration.py:
```diff
@@ -147,7 +122,7 @@ def test_target_and_harmonisee_differ(convergence_time, harmonisation_time):
-        convergence_spline=harmonisee,
+        converge_to=target,
         convergence_time=convergence_time,
         # TODO: think about convergence method
```
```suggestion
```

I think we can delete this here and deal with it in a future merge requeset

znichollscr commented on tests/integration/test_harmonise_splines_integration.py:
```diff
@@ -161,7 +136,7 @@ def test_target_and_harmonisee_differ(convergence_time, harmonisation_time):
+        convergence_spline=target,
         convergence_time=convergence_time_exp,
         # Slightly higher while we think about our numerical integration issue
         rtol=5e-3,
```
```suggestion
```

We can probably remove this now as our comparisons should be more or less exact

znichollscr commented on tests/integration/test_harmonise_splines_integration.py:
```diff
@@ -203,7 +178,7 @@ def test_more_realistic(convergence_time, harmonisation_time):
+        convergence_spline=target,
         convergence_time=convergence_time_exp,
         # Slightly higher while we think about our numerical integration issue
         rtol=1e-2,
```
```suggestion
```

We can probably remove this now as our comparisons should be more or less exact

znichollscr commented on tests/integration/test_harmonise_splines_integration.py:
```diff
@@ -189,7 +164,7 @@ def test_more_realistic(convergence_time, harmonisation_time):
-        convergence_spline=harmonisee,
+        converge_to=target,
         convergence_time=convergence_time,
         # TODO: think about convergence method
```
```suggestion
```

znichollscr commented on tests/unit/test_add_constant_to_spline.py:
```diff
@@ -0,0 +1,129 @@
+
+
+@pytest.mark.parametrize("harmonisation_time", [3.0, 3.5, 4.0])
+def test_constant_to_spline(harmonisation_time):
```
```suggestion
def test_add_constant_to_spline(harmonisation_time):
```

There's quite a lot going on in this test. Which bit exactly do you want to test?

florence-bockting replied:
Yes, it was a poor attempt to find out where the problem lies with the output of the harmonise() function. I adjusted the test now.

znichollscr commented on tests/unit/test_add_constant_to_spline.py:
```diff
@@ -0,0 +1,129 @@
+
+
+@pytest.mark.parametrize("harmonisation_time", [2015, 2016, 2017])
+def test_constant_to_spline_real_data(harmonisation_time):
```
```suggestion
def test_add_constant_to_spline_real_data(harmonisation_time):
```

There's quite a lot going on in this test. Which bit exactly do you want to test?

znichollscr commented on tests/unit/test_get_cosine_decay_spline.py:
```diff
@@ -0,0 +1,76 @@
+        in_spline=harmonised_spline_first_derivative_only, constant=diff_spline
+    )
+
+    harmonised_spline_convergence = get_cosine_decay_harmonised_spline(
```
If you're just trying to test `get_cosine_decay_harmonised_spline`, you could just start with the arguments to this function, rather than doing all the steps above to get those arguments

znichollscr reviewed (approved):
Really good overall. I've put the things I saw, but lots of them are minor so could be ignored. Have a read through then we can talk next week about any questions before we merge this in. I'll have a look at the places I noted now.

znichollscr reviewed (approved):
Really good overall. I've put the things I saw, but lots of them are minor so could be ignored. Have a read through then we can talk next week about any questions before we merge this in. I'll have a look at the places I noted now.

znichollscr committed:
Add tests of convergence module

znichollscr committed:
Pass tests

znichollscr committed:
Update integration tests

znichollscr committed:
mypy

florence-bockting committed:
update TODO and check resolved tasks

florence-bockting committed:
delete TODO as all tasks have been resolved

florence-bockting committed:
update docstring to new input parameters

florence-bockting committed:
add harmonise_splines func in harmonise() to avoid duplicate code

florence-bockting committed:
Update src/gradient_aware_harmonisation/convergence.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
change return type from SumofSplines to CosineDecaySplineHelper for code clarification.
Merge branch 'reset-fb' of https://github.com/climate-resource/gradient-aware-harmonisation into reset-fb

florence-bockting committed:
changed return type of antiderivative method for clarification

florence-bockting committed:
change order of CosineDecaySplineHelper and CosineDecaySplineHelperDerivative for logical coherence

florence-bockting committed:
update function description for better understandability

florence-bockting committed:
update comments for get-cosine-decay-harmonised-spline

florence-bockting committed:
remove higher test-tolerance value used for development as well as comments that were left over from development

florence-bockting committed:
adjusted unittest for add-constant-to-spline by specifying input parameters directly

florence-bockting committed:
fix parameter name from convergence_to to converge_to

florence-bockting committed:
Merge branch 'reset-fb' into reset-fb-zn

florence-bockting committed:
fix merging error (due to swapping CosineDecaySplineHelper and CosineDecaySplineHelperDerivative)

florence-bockting committed:
remove numpy import as unused

znichollscr committed:
Merge pull request #15 from climate-resource/reset-fb-zn

Squash bugs in cosine-decay implementation

florence-bockting committed:
add comments to tests to support test-logic

florence-bockting committed:
add pdm lock

znichollscr commented on src/gradient_aware_harmonisation/__init__.py:
```diff
@@ -57,31 +59,50 @@ def harmonise(  # noqa: PLR0913
 
-    convergence_method
+    get_harmonised_spline
         The method to use to converge back to the convergence timeseries.
```
```suggestion
        Function used to get the harmonised spline from a gradient-preserving spline and the timeseries to converge to
```

znichollscr commented on src/gradient_aware_harmonisation/convergence.py:
```diff
@@ -0,0 +1,335 @@
+            initial_time=self.initial_time,
+            final_time=self.final_time,
+            apply_to_convergence=self.apply_to_convergence,
+        )
```
```suggestion
        raise NotImplementedError
```

We'd have to be a bit careful to make this work, no?

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+    test (1-gamma) for different time values (both single value and array)
+
+    + gamma = 1-1=0 for gamma <= harmonisation_time
+    + gamma = 1-0=0 for gamma >= convergence_time
```
```suggestion
    + gamma = 1-0=1 for gamma >= convergence_time
```

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+    + gamma = 1 for gamma <= harmonisation_time
+    + gamma = 0 for gamma >= convergence_time
+    + gamma = cosine_decay(x) for harmonisation_time < gamma < convergence_time
+    """
```
```suggestion
    + gamma = 1 for time <= harmonisation_time
    + gamma = 0 for time >= convergence_time
    + gamma = cosine_decay(x) for harmonisation_time < x < convergence_time
    """
```

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+
+    + gamma = 1-1=0 for gamma <= harmonisation_time
+    + gamma = 1-0=0 for gamma >= convergence_time
+    + gamma = 1-cosine_decay(x) for harmonisation_time < gamma < convergence_time
```
```suggestion
    + gamma = 1-1=0 for time <= harmonisation_time
    + gamma = 1-0=0 for time >= convergence_time
    + gamma = 1-cosine_decay(x) for harmonisation_time < x < convergence_time
```

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+
+def test_cosine_decay_spline_derivative():
+    """
+    test derivative gamma for different time values (both single value and array)
```
```suggestion
    test derivative of gamma for different time values (both single value and array)
```

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+    test derivative gamma for different time values (both single value and array)
+
+    + gamma = 0 harmonisation_time > gamma > convergence_time
+    + gamma = cosine_decay_derivative(x) for harm_time < gamma < conv_time
```
```suggestion
    + gamma = 0 harmonisation_time > time > convergence_time
    + gamma = cosine_decay_derivative(x) for harm_time < time < conv_time
```

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+
+    + gamma = 0 harmonisation_time > gamma > convergence_time
+    + gamma = -cosine_decay_derivative(x) for harm_time < gamma < conv_time
+
```
```suggestion
    + gamma = 0 harmonisation_time > time > convergence_time
    + gamma = -cosine_decay_derivative(x) for harm_time < x < conv_time
```

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+    """
+    test derivative decay for gamma and (1-gamma)
+
+    expected_res = gamma + (-gamma) = 0
+
```
```suggestion
    expected_res = gamma + (-gamma) = 0
```

znichollscr commented on tests/integration/test_convergence_integration.py:
```diff
@@ -0,0 +1,237 @@
+    test weighted sum of two splines
+
+    gamma * spline1  + (1-gamma) * spline2 with
+    numerator = (pi * x - harmonisation_time)
```
```suggestion
    numerator = (pi * (x - harmonisation_time))
```

znichollscr commented on tests/unit/test_add_constant_to_spline.py:
```diff
@@ -0,0 +1,25 @@
+scipy = pytest.importorskip("scipy")
+
+
+def test_add_constant_to_spline():
+    x_rge = np.arange(2010, 2030, 1)
+    y_rge = np.arange(250, 270, 1)
+    constant = 2.0
+
+    spline = SplineScipy(scipy.interpolate.make_interp_spline(x=x_rge, y=y_rge))
+
+    res_spline = add_constant_to_spline(in_spline=spline, constant=constant)
+
+    np.testing.assert_allclose(res_spline(x_rge), y_rge + constant)
```
Super nice

znichollscr commented on tests/unit/test_get_cosine_decay_spline.py:
```diff
@@ -0,0 +1,76 @@
+
+@pytest.mark.parametrize("harmonisation_time", [3.0, 3.5, 4.0])
+@pytest.mark.parametrize("convergence_time", [6.0, 9.0, 10.0])
+def test_get_cosine_decay(harmonisation_time, convergence_time):
```
When I look at this, this seems to be testing the same thing as `tests/integration/test_harmonise_splines_integration.py` so we could delete this (although it doesn't hurt to have it so ultimately up to you).

znichollscr reviewed (approved):
Super. Just small things that my fresh eyes picked up, then good to merge

florence-bockting committed:
Update changelog/11.breaking.md

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update src/gradient_aware_harmonisation/__init__.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update src/gradient_aware_harmonisation/convergence.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update tests/integration/test_convergence_integration.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update tests/integration/test_convergence_integration.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update tests/integration/test_convergence_integration.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update tests/integration/test_convergence_integration.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update tests/integration/test_convergence_integration.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
Update tests/integration/test_convergence_integration.py

Co-authored-by: znichollscr <114576287+znichollscr@users.noreply.github.com>

florence-bockting committed:
reduce line length in documentation
